### PR TITLE
Add ability to set up tier prices for customer groups

### PR DESCRIPTION
**Pull request by markbeazley**
As per issue #27 this pull request adds the ability to set up tier prices per customer group.

New to Sylius development so I've tried to keep it as simple as possible, when adding a tier price, you can pick from no group or any customer group for each tier. If a variant has a tier for the customer's current group it will ignore any other tiers that don't have that group set, if there are no tiers for that group it will fallback and use the no group tiers if they are present. If the customer isn't logged in or has no group assigned it will use any tier prices with no group set.

I also ran ecs to check for style errors and it flagged a few in the existing code base which I fixed as well.

### CODE REVIEW

**Comment by mamazu** [comment]
Thank you for your contribution. This looks quite good. I am currently checking it out if the plugin works correctly.
One thing I noticed is that it is possible to create a tier price with the same quantity for the same customer group. This should not be possible for this you need to add the customer group to the unique constraint.

**Comment by markbeazley** [comment]
Yeah I was working on adding Symfony form validation on another branch, but I see you've just pushed some changes to handle the unique constraint at the Doctrine/db level, so I'll pull those in and add group to the constraint.

**Comment by markbeazley** [comment]
Just found some bugs with adding customer group to the constraint, seemed to work, but it now allows duplicates for tier prices with no group set, but does stop duplicates for tier prices with a group set.

It seems that for MySQL at least nulls are treated as if they were unique values "A UNIQUE index permits multiple NULL values for columns that can contain NULL".

Not sure on the best way to handle this.

**Comment by mamazu** [comment]
Hm if the unique constraint can only check the tierprices with customer groups then this is a good first step. The other thing is that we can and should also validate the item itself. So that there is a validator that checks if the tierprice already exists.

File: src/Traits/TierPriceableTrait.php [hunk 1]
```diff
@@ -48,33 +51,39 @@ public function getTierPrices(): array
...
     {
-        return array_filter($this->getTierPrices(), function (TierPriceInterface $tierPrice) use ($code) {
+        $channelTierPrices = array_filter($this->getTierPrices(), function (TierPriceInterface $tierPrice) use ($code) {
             $tierPriceChannel = $tierPrice->getChannel();
 
-            return $tierPriceChannel === null ? false : $tierPriceChannel->getCode() === $code;
+            return $tierPriceChannel !== null && $tierPriceChannel->getCode() === $code;
         });
+
+        return $this->filterPricesWithCustomerGroup($channelTierPrices, $customer);
```
Comment (mamazu) [inline src/Traits/TierPriceableTrait.php]: I'd suggest getting the customer group first and then filtering the array.

Comment (markbeazley) [reply src/Traits/TierPriceableTrait.php]: Just to check are you suggesting to change this function to this:

    public function getTierPricesForChannelCode(string $code, ?CustomerInterface $customer = null): array
    {
        $channelTierPrices = $this->filterPricesWithCustomerGroup($this->getTierPrices(), $customer);

        return  array_filter($channelTierPrices, function (TierPriceInterface $tierPrice) use ($code) {
            $tierPriceChannel = $tierPrice->getChannel();

            return $tierPriceChannel !== null && $tierPriceChannel->getCode() === $code;
        });
    }

Or something different? I did it the original way around as the group filtering is a little more complicated as it needs to loop through the array of tier prices once to see if there are any for that customer's group so it made sense to filter out buy channel first.

**Comment by markbeazley** [comment]
In regards to validation, I got the UniqueEntity constraint to work by setting its ignoreNull option to false, however the UniqueEntity constraint can only check against entities already persisted to the database, so doesn't catch the case where the user tries to insert two identical quantities at the same time, I'm currently working on a custom constraint that can check against other submitted ones.

File: src/Traits/TierPriceableTrait.php (hunk 1 above)
Comment (mamazu) [reply src/Traits/TierPriceableTrait.php]: One thing I also thought is to save the prices in the database by channel key so that when reading it we only need to access the right key and then we'd have it.

**Comment by mamazu** [comment]
When users try to add stuff at the same time then one will go first and the other one will go second and get the validation error. But if you can get the validator to work to prevent it go ahead. :)

**Comment by markbeazley** [comment]
When I was testing it it seems it only validates all objects once and then tries to insert them. So if you try and add two identical quantities with the same customer group you get an exception thrown from Doctrine as the database unique constraint fails when it tries to insert the records. If you try and add two identical quantities with no group set (so null) it successfully inserts both as the unique constraint in MySQL (and most over database servers) treat 2 nulls as different values.

**Comment by mamazu** [comment]
I will have a look at it. Can you try to fix the conflicts with the master branch in the meantime?

**Comment by mamazu** [comment]
I added validation. Before you continue to work on the branch please pull first.

**Comment by markbeazley** [comment]
Sorry for the delay, I tested out the validation you did and it seems to have the same issue as the default UniqueEntity one, it only checks against against the entries in the database already this can be seen by starting with a variation with no tier prices, click add, set the quantity to 10, price to any vcalue and leave customer group set to none, then click add and set the second new one to the same quantity and a customer group of none as well and now click save, it will successfully save the duplicate entries. It will fail with an exception if you do the same thing but this time setting the customer group to the same group as this will cause the database level unique constraint to fail.

I've changed the validator to get the full list of objects from the form, then using reflection code similar to how the default UniqueEntity class works to do the comparison.

**Comment by mamazu** [comment]
If it works now you can rebase it onto the current master and we will merge it.

**Comment by markbeazley** [comment]
I've merged in the changes from master.

I'm also wondering if the validation.xml file is in the wrong place, but I've not done much Sylius plugin work before so not sure, but at the moment its in tests/Application/config/validator folder, this works when you are testing the plugin with the copy of sylius inside the tests/Application folder but would it get picked up when the plugin is installed via composer in a sylius store? I've tested this on a separate branch [here](https://github.com/markbeazley/SyliusTierpricePlugin/commit/8c01c14c2905bfc71713f5b374de6857ccd0112f) and it is still working when testing the plugin on its own.

**Comment by markbeazley** [comment]
Was just testing this in our main sylius store, and noticed an odd column naming issue which is causing the <unique-constraint columns="qty,channel_id,productVariant_id,customerGroup_id" name="no_duplicate_prices"/> to throw an exception when doctrine tries to process it.

Basically when I run `(cd tests/Application && bin/console doctrine:schema:update --dump-sql -e dev)` from my clone of the plug in repo using MySql for the database, the productVariant and csutomerGroup columns are being generated as productVariant_id and customerGroup_id which match the values in the unique constraint, but not the naming scheme doctrine seems to be using for the rest of the database (all lower case _ used as a separator). 

When I run the equivalent command in our main sylius store repository using the latest code from my branch which uses PostgreSQL for the database, those two columns are formatted as I would expect. product_variant_id and customer_group_id, but as the unique constraint is referencing them as product_variant_id, customer_group_id I just get the error `There is no column with name 'productVariant_id' on table 'brille24_tierprice'.` unless I modify the constraint to use the alternative naming style.

**Comment by markbeazley** [comment]
Right I've done some more digging, seems our main sylius shop is setting Doctrine to use the underscore naming strategy. Looking at the sylius core *.orm.xml files, they all manually define a column name for each property and they seem to be following the underscore strategy.

I've added a commit [here](https://github.com/markbeazley/SyliusTierpricePlugin/commit/8dbd564e39347e4ef773d93e49d7b66828f95674) that makes TierPrice.orm.xml follow the example of the core Sylius files.

**Comment by mamazu** [comment]
About the validation: I think we need to mention this in the read me to copy the validation into the project folder when installing the plugin.

About the key contraints, you are right this is can cause problems if we use a different naming strategy so adding the name of the join column is a good idea.

**Comment by markbeazley** [comment]
With the validation I tested moving it from tests/Application/config/validator/validation.xml to src/Resources/config/validation/validation.xml and it seems to work both testing the plugin directly on its own, and inside an existing install without needing the user to copy the validation. Being new to Sylius development is there any reason or advantage to require the user to copy the validation themselves?

With the key constraints I can merge the branch I made with the join column names in if you are ok with the style I've picked: product_variant_id and customer_group_id. These seem to match the naming style Sylius are using for the core entities.

**Comment by mamazu** [comment]
If it works out of the box without copying the validation over then even better, then we don't need to specify it. There is no real benefit of it.

Other than that this looks good to go.

**Comment by mamazu** [comment]
@kortwotze What do you think about this?

**Comment by markbeazley** [comment]
I've merged into my fork the move of the validation config and the change to specify column names to prevent issues with different doctrine naming strategies.

File: src/Services/TierPriceFinder.php [hunk 2]
```diff
@@ -33,16 +34,22 @@ public function __construct(TierPriceRepositoryInterface $tierPriceRepository)
...
         ChannelInterface $channel,
+        ?CustomerInterface $customer,
         int $quantity
     ): ?TierPriceInterface {
-        $possibleTierPrices = $this->tierPriceRepository->getSortedTierPrices($tierPriceableEntity, $channel);
+        $group = null;
+        if ($customer instanceof CustomerInterface) {
+            $group = $customer->getGroup();
+        }
+        $possibleTierPrices = $this->tierPriceRepository->getSortedTierPrices($tierPriceableEntity, $channel, $group);
```
Comment (kortwotze) [inline src/Services/TierPriceFinder.php]: This is exactly what I was talking about in the other comment. ;)

File: src/Traits/TierPriceableTrait.php (hunk 1 above)
Comment (kortwotze) [reply src/Traits/TierPriceableTrait.php]: I like that - having the channel reference on the entity makes a lot of sense, if its channelized anyway :D

File: src/Resources/config/doctrine/TierPrice.orm.xml [hunk 3]
```diff
@@ -20,8 +20,10 @@
 
         <many-to-one target-entity="Sylius\Component\Product\Model\ProductVariantInterface" field="productVariant" inversed-by="tierPrices"/>
 
+	<many-to-one target-entity="Sylius\Component\Customer\Model\CustomerGroupInterface" field="customerGroup" />
```
[the remark this answers is not in the document]
Comment (markbeazley) [reply src/Resources/config/doctrine/TierPrice.orm.xml]: Yeah, I had that on a separate branch which I didn't merge before I left work on Friday, merged now.

File: src/Repository/TierPriceRepository.php [hunk 4]
```diff
@@ -17,15 +17,36 @@
...
     {
-        return $this->findBy(['productVariant' => $productVariant, 'channel' => $channel], ['qty' => 'ASC']);
+        if ($customerGroup instanceof CustomerGroupInterface) {
+            $prices = $this->findBy(['productVariant' => $productVariant, 'channel' => $channel, 'customerGroup' => $customerGroup], ['qty' => 'ASC']);
+            if (count($prices) > 0) {
+                return $prices;
+            }
+        }
+
+        return $this->findBy(['productVariant' => $productVariant, 'channel' => $channel, 'customerGroup' => null], ['qty' => 'ASC']);
```
[the remark this answers is not in the document]
Comment (markbeazley) [reply src/Repository/TierPriceRepository.php]: It might not be clear what is happening here, I've designed it so that if a customer is a member of a group and a variant has a tier price for their group it should only use the tier prices for their group, if there are no tier prices for their group but there are tier prices for no group it should fallback to them. I need to query the database to find out if there are any tier prices for their specific group, if there are I return early, other wise I need to run the query for the no group tier prices.

File: src/Repository/TierPriceRepository.php [hunk 5]
```diff
@@ -17,15 +17,36 @@
...
+use Sylius\Component\Customer\Model\CustomerGroupInterface;
 
 class TierPriceRepository extends EntityRepository implements TierPriceRepositoryInterface
 {
     /** {@inheritdoc}
      *
      * @return TierPriceInterface[]
      */
-    public function getSortedTierPrices(TierPriceableInterface $productVariant, ChannelInterface $channel): array
+    public function getSortedTierPrices(TierPriceableInterface $productVariant, ChannelInterface $channel, ?CustomerGroupInterface $customerGroup): array
```
[the remark this answers is not in the document]
Comment (markbeazley) [reply src/Repository/TierPriceRepository.php]: I can do this, didn't do it that way originally as it was only used in the plugin in one other place but forgot people might be using it outside the plugin.

[the remark this answers is not in the document]
Comment (mamazu) [reply src/Repository/TierPriceRepository.php]: Or just add a default value to this function which would have the same effect.

File: src/Repository/TierPriceRepository.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (mamazu) [reply src/Repository/TierPriceRepository.php]: Yeah we probably need to document this behaviour.

File: src/Repository/TierPriceRepository.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (markbeazley) [reply src/Repository/TierPriceRepository.php]: I've gone with a default value as that is the method I've used elsewhere and results in less code.

**Comment by markbeazley** [comment]
By adding the column names to the TierPrice.orm.xml depending on the naming strategy the user is using in their Sylius install the auto generated database migration would potentially result in data loss for example `ALTER TABLE brille24_tierprice ADD product_variant_id INT DEFAULT NULL, ADD customer_group_id INT DEFAULT NULL, DROP productVariant_id, DROP customerGroup_id;`  was generated on my development system would drop the productVariant_id thus losing the productVarient set on any existing tier prices.

Not sure on the best course of action here, Sylus by default doesn't specify any specific naming strategy to Doctrine so it will use the default strategy (productVarient_id) so it might make sense to use that strategy in TierPrice.orm.xml instead as it might effect less people that way, but would still effect anyone who is using any naming strategy other than the Doctrine default so would still need to be documented for users somewhere. 

The fix  from the user perspective would be to manually edit the migration so that the query 

`ALTER TABLE brille24_tierprice ADD product_variant_id INT DEFAULT NULL, ADD customer_group_id INT DEFAULT NULL, DROP productVariant_id, DROP customerGroup_id;`

Is split in 3 like so:

`ALTER TABLE brille24_tierprice ADD product_variant_id INT DEFAULT NULL, ADD customer_group_id INT DEFAULT NULL;`
`UPDATE brille24_tierprice set product_variant_id = productVariant_id;`
`ALTER TABLE brille24_tierprice DROP productVariant_id, DROP customerGroup_id;`

Obviously the queries would have to be different based on the naming strategy and database the user is using.

File: src/Repository/TierPriceRepository.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (kortwotze) [reply src/Repository/TierPriceRepository.php]: Well, I got that behaviour - I just think the implementation can be clarified a little ;) Both queries use the same set of parameters (one with `null` as value, one with a customer group. If you check [this comment](https://github.com/Brille24/SyliusTierpricePlugin/pull/37#discussion_r334374581), you see what I mean. You can put the query parameters together and then use only one return statement to achieve the desired result.

File: src/Repository/TierPriceRepository.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (kortwotze) [reply src/Repository/TierPriceRepository.php]: :+1:

File: src/Resources/config/doctrine/TierPrice.orm.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (kortwotze) [reply src/Resources/config/doctrine/TierPrice.orm.xml]: Seen. :+1:

I still have the feeling the whole naming-scheme of doctrine is not properly defined in Sylius and thus in this plugin.

I am leaning to forcing the column names to meet Sylius' expectations until they have defined it nicer.

File: src/Repository/TierPriceRepository.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (markbeazley) [reply src/Repository/TierPriceRepository.php]: I think I would still need to run the two queries under certain code paths for example If I get no results for the query on line 31 I need to run it again with null passed to customer group to get any potential fallback tier prices. As an example with this set up:
image a customer with any group other than Wholesale would need to get the tier prices with no customer group set. 

I don't think I can get that behaviour with a single query unless I changed it to get entries where customerGroup is null or equal to the current customers group then did the filtering in PHP similar to TierPriceableTrait::filterPricesWithCustomerGroup().

[the remark this answers is not in the document]
Comment (markbeazley) [reply src/Repository/TierPriceRepository.php]: I've removed the early return and extracted the query parameters to an array on another branch [here](https://github.com/markbeazley/SyliusTierpricePlugin/commit/67c4ed418573cedca1def7f27f45200afbcd5a25), I still have to make the second query if the one for a specific customer group returns no tier prices however.

**Comment by markbeazley** [comment]
Hi just checking in to see if it would be possible to get an update on anything else I would need to do to get this merged, we are waiting on this feature for some sites I am currently working on and whilst I could just install my forked copy it would be cleaner if I could get this merged.

**Comment by mamazu** [comment]
I am all for it. So we should merge it. What do you think @kortwotze

File: src/Repository/TierPriceRepository.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (mamazu) [reply src/Repository/TierPriceRepository.php]: Yeah this is currently the best bet for it. I think we should keep it.

File: src/Resources/translations/messages.de.yml [hunk 6]
```diff
@@ -1,6 +1,4 @@
 brille24_tier_price:
     ui:
         tier_prices: Staffelpreise
-    form:
```
Comment (mamazu) [inline src/Resources/translations/messages.de.yml]: Why did you remove the form validation?

File: src/Validator/TierPriceUniqueValidator.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
+<?php
```
Comment (mamazu) [inline src/Validator/TierPriceUniqueValidator.php]: Why do we need to override the unique entity validator?

File: src/Resources/translations/messages.de.yml (hunk 6 above)
Comment (markbeazley) [reply src/Resources/translations/messages.de.yml]: As per the [Symfony Docs](https://symfony.com/doc/current/validation/translations.html) it seems form validation messages need to be in validators.*.yml files so I moved them there. I wasn't seeing the translated message with them in messages.*.yml even after a cache clear.

File: src/Validator/TierPriceUniqueValidator.php (hunk 7 above)
Comment (markbeazley) [reply src/Validator/TierPriceUniqueValidator.php]: With the built in Unique entity validator, and the version you added, there is an edge case when the user tries to add the same quantity and group choice in one action. Both the Unique entity validator and your one only check against ones persisted to the database, and the check only happens once before any are submitted, my version checks against the data submitted in the form, which catches all instances of duplicate versions.

File: src/Resources/translations/messages.de.yml (hunk 6 above)
Comment (mamazu) [reply src/Resources/translations/messages.de.yml]: Ah good call. Thanks for spotting.

File: src/Validator/TierPriceUniqueValidator.php (hunk 7 above)
Comment (mamazu) [reply src/Validator/TierPriceUniqueValidator.php]: And we can't reuse part of the default validator and just decorate it?

Comment (markbeazley) [reply src/Validator/TierPriceUniqueValidator.php]: We might be able to, not really up to speed with decorating, not sure what else I could do other than extend the UniqueEntityValidator, then change validate() to run parent::validate() before running my code which might lead to some double checking.

Comment (mamazu) [reply src/Validator/TierPriceUniqueValidator.php]: When decorating the class you're class has the other class as an argument in the constructor and then you call the method that way. This makes it easier to preserve the original logic.

Comment (markbeazley) [reply src/Validator/TierPriceUniqueValidator.php]: If I use decorating is it possible to just decorate it for this one field or would it be global throughout the app? My worry would be if it was global it might break code elsewhere that relies on it only checking already persisted entities.

Comment (mamazu) [reply src/Validator/TierPriceUniqueValidator.php]: True then inheritance might be the better way to go.

Comment (markbeazley) [reply src/Validator/TierPriceUniqueValidator.php]: Been having a play around with this the last few days, I can't just extend the UniqueEntityValidator as it requires a different constraint to TierPriceUniqueValidator. I can achieve much the same by adding a second constraint using UniqueEntity, but this results in both validators flagging the same duplicates multiple times.

Comment (mamazu) [reply src/Validator/TierPriceUniqueValidator.php]: Then I think this is the best approach.

**Comment by mamazu** [comment]
Thanks for your work, I have tested it in our stack and it works. So merged and tagged it is. :tada:

**Comment by markbeazley** [comment]
@mamazu Brilliant thanks, should I document the potential data loss issue when upgrading that I out lined [here](https://github.com/Brille24/SyliusTierpricePlugin/pull/37#issuecomment-541732103) and if so where would be the best place?

**Comment by mamazu** [comment]
That would be good. The best place would be in the `UPGRADE.md`.
